src/exclusion/object_filter.hpp

```diff
@@ -22,35 +24,131 @@
+
+        ~trie_node() = default;
+        trie_node(const trie_node&) = delete;
+        trie_node(trie_node&&) = default;
+        trie_node& operator=(const trie_node&) = delete;
+        trie_node& operator=(trie_node&&) = default;
+
+        struct comp {
+            bool operator()(const trie_node& n, std::string_view v) {
+                return std::string_view{n.path_component} < v;
+            }
+            bool operator()(std::string_view v, const trie_node& n) {
+                return v < std::string_view{n.path_component};
+            }
+        };
+
+        [[nodiscard]] trie_node const *get_child(std::string_view key) const {
+            auto [beg, end] = std::equal_range(
+                children.begin(), children.end(), key, comp{});
+            if (beg != end) {
+                return &*beg;
+            }
+            return nullptr;
+        }
+
+        trie_node& get_or_create_child(std::string_view key, bool& is_new) {
+            auto [beg, end] = std::equal_range(
+                children.begin(), children.end(), key, comp{});
+            if (beg != end) {
+                is_new = false;
+                return *beg;
+            }
+
+            auto ins_it = children.emplace(end, std::string{key});
+            is_new = true;
+            return *ins_it;
+        }
+
+        [[nodiscard]] bool is_terminal() const {
+            return children.empty();
+        }
     };
+    static_assert(std::is_move_assignable<trie_node>::value);
 
-    explicit path_trie(std::shared_ptr<trie_node> node_) : root(std::move(node_)) {}
+    trie_node root{""};
+    // we allow adding the root to the trie (matching everything)
+    // so we use this flag to distinguish the two cases (empty vs everything)
+    bool is_empty{true}; 
 
-    std::string_view get_stored_string(std::string_view str);
+public:
+    class traverser {
+        trie_node const *const cur_node{};
+    public:    
+        enum class state {
+            NOT_FOUND,
+            FOUND,
+            GLUE
+        };
+
+        explicit traverser(trie_node const *root) : cur_node{root} {}
+        [[nodiscard]] traverser descend(std::string_view next_key) const {
+            if (get_state() != state::GLUE) {
+                // once FOUND/NOT_FOUND, as we descend we keep the state
+                return *this;
+            }
+            return traverser{cur_node->get_child(next_key)};
+        }
+        [[nodiscard]] state get_state() const {
+            if (cur_node == nullptr) {
+                return state::NOT_FOUND;
+            }
+            return cur_node->is_terminal() ? state::FOUND : state::GLUE;
+        }
+    };
+
+    template <typename StringType,
+        typename = std::enable_if<std::is_constructible<std::string, StringType>::value>>
+    void insert(const std::vector<StringType> &path)
+    {
+        is_empty = false;
+
+        trie_node *cur = &root;
+        bool is_new;
```

Comment:
If the path is empty this will be used uninitialised.

Reply:
yeah, better to use the pair...